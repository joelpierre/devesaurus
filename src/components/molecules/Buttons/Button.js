import React from 'react';
import { Link } from 'gatsby';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Button.module.scss';

const Button = ({ text, behavior, theme, link, size, action }) => {
  let button;
  switch (behavior) {
    case 'router':
      button = (
        <Link
          data-test="component-button"
          to={link}
          className={classNames([
            styles.btn,
            styles[`btn--${size}`],
            styles[`btn--${theme}`],
          ])}
        >
          {text}
        </Link>
      );
      break;

    case 'anchor':
      button = (
        <a
          data-test="component-button"
          href={link}
          className={classNames([
            styles.btn,
            styles[`btn--${size}`],
            styles[`btn--${theme}`],
          ])}
          target="_blank"
          rel="noopener noreferrer nofollow"
        >
          {text}
        </a>
      );
      break;

    case 'action':
      button = (
        <button
          data-test="component-button"
          type="button"
          className={classNames([
            styles.btn,
            styles[`btn--${size}`],
            styles[`btn--${theme}`],
          ])}
          onClick={action.func}
        >
          {text}
        </button>
      );
      break;
  }

  return <>{button}</>;
};

Button.defaultProps = {
  theme: 'brand',
  behavior: 'router',
  size: 'md',
  action: null,
};

Button.propTypes = {
  action: PropTypes.func,
  text: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  theme: PropTypes.oneOf([
    'brand',
    'alpha',
    'tint-alpha',
    'tint-beta',
    'tint-psi',
    'tint-omega',
  ]),
  behavior: PropTypes.oneOf(['router', 'anchor', 'action']),
  size: PropTypes.oneOf(['sm', 'md', 'lg', 'xl', 'full']),
};

export default Button;
