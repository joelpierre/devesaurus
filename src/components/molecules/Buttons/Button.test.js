import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import {
  checkProps,
  findByTestAttr,
  matchSnapshot,
} from '../../../utils/test-utilities';

import Button from './Button';

const defaultProps = {
  text: 'test',
  link: '#',
};

/**
 * Factory function to create a ShallowWrapper for the Button component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
  const setupProps = { ...defaultProps, ...props };
  return shallow(<Button {...setupProps} />);
};

describe('<Button/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('renders the Button Component with router behaviour', () => {
    const component = findByTestAttr(wrapper, 'component-button');
    expect(component.length).toBe(1);
  });

  it('renders the Button Component with anchor behaviour', () => {
    wrapper = setup({
      ...defaultProps,
      behavior: 'anchor',
    });
    const component = findByTestAttr(wrapper, 'component-button');
    expect(component.length).toBe(1);
  });

  it('renders the Button Component with action behaviour', () => {
    wrapper = setup({
      ...defaultProps,
      behavior: 'action',
      action: jest.fn(),
    });
    const component = findByTestAttr(wrapper, 'component-button');
    expect(component.length).toBe(1);
  });

  it('should render with correct props', () => {
    checkProps(Button, defaultProps);
  });

  it('should match snapshot', () => {
    matchSnapshot(wrapper);
  });
});
