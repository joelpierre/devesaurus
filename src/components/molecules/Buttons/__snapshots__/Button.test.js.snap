// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should match snapshot 1`] = `
<Fragment>
  <mockConstructor
    className="btn btn--md btn--brand"
    data-test="component-button"
    to="#"
  >
    test
  </mockConstructor>
</Fragment>
`;
